streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -252,6 +288,111 @@ public void buildAndOptimizeTopology() {
         }
     }
 
+    private void maybePerformOptimizations(final Properties props) {
+
+        if (props != null && props.getProperty(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE)) {
+            LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
+            maybeOptimizeRepartitionOperations();
+        }
+    }
+
+    private void maybeOptimizeRepartitionOperations() {
+        StreamsGraphNode optimizedSingleRepartition = null;
+
+        for (final Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> streamsGraphNodeSetEntry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
+
+            StreamsGraphNode keyChangingNode = streamsGraphNodeSetEntry.getKey();
+
+            if (streamsGraphNodeSetEntry.getValue().isEmpty()) {
+                continue;
+            }
+
+            final StreamSourceNode parentSourceNode = (StreamSourceNode) findParentNodeMatching(keyChangingNode, n -> n instanceof StreamSourceNode);
+
+            if (parentSourceNode == null) {
+                throw new IllegalStateException(String.format("Can't find parent source node for %s ", keyChangingNode));
+            }
+
+            optimizedSingleRepartition = createRepartitionNode(keyChangingNode.nodeName(),
+                                                               parentSourceNode.keySerde(),
+                                                               parentSourceNode.valueSerde());
+
+            // re-use parent id to make sure the single repartition graph node is evaluated before downstream nodes
+            optimizedSingleRepartition.setId(parentSourceNode.id());
+
+            for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : streamsGraphNodeSetEntry.getValue()) {
+
+                final StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNodes().contains(keyChangingNode));
+
+                LOG.debug("Found the node downstream of key-changer {}", keyChangingNodeChild);
+
+                // need to add children of key-changing node as children of optimized repartition
+                // in order to process records from re-partitioning
+                optimizedSingleRepartition.addChildNode(keyChangingNodeChild);
+
+                LOG.debug("Removing {} from {}  children {}", keyChangingNodeChild, keyChangingNode, keyChangingNode.children());
+                // now remove children from key-changing node
+                keyChangingNode.removeChild(keyChangingNodeChild);
+
+                // now need to get children of repartition node so we can remove repartition node
+                final Collection<StreamsGraphNode> repartitionNodeToBeReplacedChildren = repartitionNodeToBeReplaced.children();
+
+                final StreamsGraphNode parentOfRepartitionNodeToBeReplaced = findParentNodeMatching(repartitionNodeToBeReplaced, n -> repartitionNodeToBeReplaced.parentNodes().contains(n));
```

Comment:
I think this says, "we are looking for a parent node that is among my parent nodes".

It seems like we can generalize this logic by not assuming there's a single parent node at all.

Reply:
Unless I'm missing something the `findParenNodeMatching` already takes into account that there is possibly more than one parent node.  Plus this method traverses "up" the graph until it finds a match so the returned node could be a grandparent or higher, not just searching the immediate collection of parent graph nodes.

Reply:
Sure, but the predicate restricts the results to just the parents of `repartitionNodeToBeReplaced`, this method call just serves to get the last parent node.

It just seems like a roundabout way to do `repartitionNodeToBeReplaced.parentNodes().get(0)` or  `repartitionNodeToBeReplaced.parentNodes().iterator().next()` or similar.

It is true that `repartitionNodeToBeReplaced` accounts for the fact that `parents` is a collection, but the logic still just re-roots the children to one of the parents, so it really only produces sensible results when there is exactly one parent.

This is what I meant by generalizing:
```java
for (final StreamsGraphNode parentNode : repartitionNodeToBeReplaced.parentNodes()) {
                parentNode.removeChild(repartitionNodeToBeReplaced);
                for (final StreamsGraphNode childNode : repartitionNodeToBeReplaced.children()) {
                    parentNode.addChildNode(childNode);
                }
}
```
which works perfectly fine when there's just one parent, but also works when there are multiple parents, although I'm not sure if the results are sensible either.

It might be clearer and safer to just say `final StreamsGraphNode parentOfRepartitionNodeToBeReplaced = getOnly(repartitionNodeToBeReplaced.parentNodes())`